core/modules/cart/store/actions/mergeActions.ts

```diff
@@ -0,0 +1,196 @@
+import * as types from '@vue-storefront/core/modules/cart/store/mutation-types'
+import { Logger } from '@vue-storefront/core/lib/logger'
+import config from 'config'
+import EventBus from '@vue-storefront/core/compatibility/plugins/event-bus'
+import { CartService } from '@vue-storefront/core/data-resolver'
+import {
+  productsEquals,
+  createDiffLog,
+  notifications,
+  createCartItemForUpdate
+} from '@vue-storefront/core/modules/cart/helpers'
+import CartItem from '@vue-storefront/core/modules/cart/types/CartItem';
+
+const mergeActions = {
+  async updateClientItem ({ dispatch }, { clientItem, serverItem }) {
+    const cartItem = clientItem === null ? await dispatch('getItem', serverItem) : clientItem
+
+    if (!cartItem || typeof serverItem.item_id === 'undefined') return
+
+    const product = {
+      server_item_id: serverItem.item_id,
+      sku: cartItem.sku,
+      server_cart_id: serverItem.quote_id,
+      prev_qty: cartItem.qty,
+      product_option: serverItem.product_option,
+      type_id: serverItem.product_type
+    }
+
+    await dispatch('updateItem', { product })
+    EventBus.$emit('cart-after-itemchanged', { item: cartItem })
+  },
+  async updateServerItem ({ getters, rootGetters, commit, dispatch }, { clientItem, serverItem, updateIds }) {
+    const diffLog = createDiffLog()
+    const cartItem = createCartItemForUpdate(clientItem, serverItem, updateIds)
+    const event = await CartService.updateItem(getters.getCartToken, cartItem)
+    const isUpdateSuccess = event.resultCode === 200
+    Logger.debug('Cart item server sync' + event, 'cart')()
+    diffLog.pushServerResponse({ status: event.resultCode, sku: clientItem.sku, result: event })
+
+    if (!isUpdateSuccess && !serverItem) {
+      commit(types.CART_DEL_ITEM, { product: clientItem, removeByParentSku: false })
+      return diffLog
+    }
+
+    if (!isUpdateSuccess && clientItem.item_id) {
+      await dispatch('restoreQuantity', { cartItem, clientItem })
+      return diffLog
+    }
+
+    if (!isUpdateSuccess) {
+      Logger.warn('Removing product from cart', 'cart', clientItem)()
+      commit(types.CART_DEL_NON_CONFIRMED_ITEM, { product: clientItem })
+      return diffLog
+    }
+
+    if (!rootGetters['checkout/isUserInCheckout']) {
+      const isThisNewItemAddedToTheCart = (!clientItem || !clientItem.server_item_id)
+      diffLog.pushNotification(
+        isThisNewItemAddedToTheCart ? notifications.productAddedToCart : notifications.productQuantityUpdated
+      )
+    }
+
+    await dispatch('updateClientItem', { clientItem, serverItem: event.result })
+
+    return diffLog
+  },
+  async synchronizeServerItem ({ dispatch }, { serverItem, clientItem, forceClientState, dryRun }) {
```

Comment:
it's pretty hard to get the difference between `synchronizeServerItem` vs `mergeServerItem` - not sure if this separate action does more good or bad for hte overall complexity to be honest. Just raising this as an suggestion not sure how it should be finally done. For me it's even OK as it's now

Reply:
@pkarw I understand... basically this action was created only for `mergeClientItem` and I can call it even `syncClientItemWithServerItem` or something like that. I have to consider that & I'll do some changes just before the merge